File: packages/sdk/src/dialog.ts
import { host } from "./host";

/**
 * Native OS dialog API namespace for file and directory selection.
 *
 * Provides access to native OS file picker dialogs. These dialogs are
 * rendered by the operating system and appear as standard system windows.
 *
 * @example
 * ```typescript
 * import { dialog } from "@hmcs/sdk";
 *
 * // Pick a single file
 * const file = await dialog.pickFile({ filters: [{ name: "Images", extensions: ["png", "jpg"] }] });
 *
 * // Pick multiple files
 * const files = await dialog.pickFiles();
 *
 * // Pick a folder
 * const folder = await dialog.pickFolder();
 * ```
 */
export namespace dialog {
  /**
   * A file type filter for native file dialogs.
   *
   * @example
   * ```typescript
   * const imageFilter: dialog.FileFilter = {
   *   name: "Images",
   *   extensions: ["png", "jpg", "jpeg", "gif"]
   * };
   * ```
   */
  export interface FileFilter {
    /** Display name for the filter (e.g. "Images"). */
    name: string;
    /** File extensions without leading dot (e.g. `["png", "jpg"]`). */
    extensions: string[];
  }

  /** Options for file picker dialogs. */
  export interface PickFileOptions {
    /** Dialog window title. */
    title?: string;
    /** Initial directory to open the dialog in. */
    defaultPath?: string;
    /** File type filters shown in the dialog. */
    filters?: FileFilter[];
  }

  /** Options for folder picker dialog. */
  export interface PickFolderOptions {
    /** Dialog window title. */
    title?: string;
    /** Initial directory to open the dialog in. */
    defaultPath?: string;
  }

  /**
   * Opens a native OS single-file picker dialog.
   *
   * Displays the platform's standard file selection dialog. Returns the
   * selected file path as a string, or `null` if the user cancels.
   *
   * @param options - Optional dialog configuration
   * @returns The selected file path, or `null` if cancelled
   *
   * @example
   * ```typescript
   * const file = await dialog.pickFile({
   *   title: "Select a VRM model",
   *   filters: [{ name: "VRM", extensions: ["vrm"] }],
   *   defaultPath: "/Users/me/models"
   * });
   * if (file) {
   *   console.log(`Selected: ${file}`);
   * }
   * ```
   */
  export async function pickFile(options?: PickFileOptions): Promise<string | null> {
    const response = await host.post(host.createUrl("dialog/pick-file"), options);
    const body = (await response.json()) as { path: string | null };
    return body.path;
  }

  /**
   * Opens a native OS multi-file picker dialog.
   *
   * Displays the platform's standard multi-file selection dialog. Returns
   * an array of selected file paths, or an empty array if the user cancels.
   *
   * @param options - Optional dialog configuration
   * @returns Array of selected file paths (empty if cancelled)
   *
   * @example
   * ```typescript
   * const files = await dialog.pickFiles({
   *   title: "Select images",
   *   filters: [{ name: "Images", extensions: ["png", "jpg", "jpeg"] }]
   * });
   * for (const file of files) {
   *   console.log(`Selected: ${file}`);
   * }
   * ```
   */
  export async function pickFiles(options?: PickFileOptions): Promise<string[]> {
    const response = await host.post(host.createUrl("dialog/pick-files"), options);
    const body = (await response.json()) as { paths: string[] };
    return body.paths;
  }

  /**
   * Opens a native OS directory picker dialog.
   *
   * Displays the platform's standard folder selection dialog. Returns the
   * selected directory path as a string, or `null` if the user cancels.
   *
   * @param options - Optional dialog configuration
   * @returns The selected directory path, or `null` if cancelled
   *
   * @example
   * ```typescript
   * const folder = await dialog.pickFolder();
   * if (folder) {
   *   console.log(`User chose: ${folder}`);
   * }
   * ```
   */
  export async function pickFolder(options?: PickFolderOptions): Promise<string | null> {
    const response = await host.post(host.createUrl("dialog/pick-folder"), options);
    const body = (await response.json()) as { path: string | null };
    return body.path;
  }
}

File: engine/crates/homunculus_http_server/src/route/dialog.rs
//! `/dialog` provides native OS dialog operations.

use axum::Json;
use serde::{Deserialize, Serialize};
use utoipa::ToSchema;

/// A file type filter for native file dialogs.
#[derive(Serialize, Deserialize, Debug, Clone, PartialEq, ToSchema)]
#[serde(rename_all = "camelCase")]
pub struct FileFilter {
    /// Display name for the filter (e.g. "Images").
    pub name: String,
    /// File extensions without leading dot (e.g. `["png", "jpg"]`).
    pub extensions: Vec<String>,
}

/// Request body for file picker dialogs.
#[derive(Serialize, Deserialize, Debug, Clone, Default, ToSchema)]
#[serde(rename_all = "camelCase")]
pub struct PickFileRequest {
    /// Dialog window title. Defaults to `"Select File"` or `"Select Files"`.
    pub title: Option<String>,
    /// File type filters shown in the dialog.
    pub filters: Option<Vec<FileFilter>>,
    /// Initial directory to open the dialog in.
    pub default_path: Option<String>,
}

/// Request body for folder picker dialog.
#[derive(Serialize, Deserialize, Debug, Clone, Default, ToSchema)]
#[serde(rename_all = "camelCase")]
pub struct PickFolderRequest {
    /// Dialog window title. Defaults to `"Select Directory"`.
    pub title: Option<String>,
    /// Initial directory to open the dialog in.
    pub default_path: Option<String>,
}

/// Response from the single file picker dialog.
#[derive(Serialize, Deserialize, Debug, Clone, ToSchema)]
#[serde(rename_all = "camelCase")]
pub struct PickFileResponse {
    /// The selected file path, or `null` if the user cancelled.
    pub path: Option<String>,
}

/// Response from the multi-file picker dialog.
#[derive(Serialize, Deserialize, Debug, Clone, ToSchema)]
#[serde(rename_all = "camelCase")]
pub struct PickFilesResponse {
    /// The selected file paths. Empty if the user cancelled.
    pub paths: Vec<String>,
}

/// Response from the folder picker dialog.
#[derive(Serialize, Deserialize, Debug, Clone, ToSchema)]
#[serde(rename_all = "camelCase")]
pub struct PickFolderResponse {
    /// The selected directory path, or `null` if the user cancelled.
    pub path: Option<String>,
}

/// Open a native OS directory picker dialog.
///
/// Returns the selected directory path. If the user cancels the dialog,
/// `path` will be `null`. Optionally accepts a title and default directory.
#[utoipa::path(
    post,
    path = "/pick-folder",
    tag = "dialog",
    request_body(content = Option<PickFolderRequest>, content_type = "application/json"),
    responses(
        (status = 200, description = "Folder selection result", body = PickFolderResponse),
    ),
)]
pub async fn pick_folder(body: Option<Json<PickFolderRequest>>) -> Json<PickFolderResponse> {
    let req = body.map(|Json(r)| r).unwrap_or_default();
    let dialog = build_file_dialog(
        req.title.as_deref(),
        "Select Directory",
        None,
        req.default_path.as_deref(),
    );
    let handle = dialog.pick_folder().await;
    let path = handle.map(|h| h.path().to_string_lossy().to_string());
    Json(PickFolderResponse { path })
}

/// Open a native OS single-file picker dialog.
///
/// Returns the selected file path. If the user cancels the dialog,
/// `path` will be `null`. Optionally accepts filters and a default directory.
#[utoipa::path(
    post,
    path = "/pick-file",
    tag = "dialog",
    request_body(content = Option<PickFileRequest>, content_type = "application/json"),
    responses(
        (status = 200, description = "File selection result", body = PickFileResponse),
    ),
)]
pub async fn pick_file(body: Option<Json<PickFileRequest>>) -> Json<PickFileResponse> {
    let req = body.map(|Json(r)| r).unwrap_or_default();
    let dialog = build_file_dialog(
        req.title.as_deref(),
        "Select File",
        req.filters.as_deref(),
        req.default_path.as_deref(),
    );
    let handle = dialog.pick_file().await;
    let path = handle.map(|h| h.path().to_string_lossy().to_string());
    Json(PickFileResponse { path })
}

/// Open a native OS multi-file picker dialog.
///
/// Returns the selected file paths. If the user cancels the dialog,
/// `paths` will be an empty array. Optionally accepts filters and a default directory.
#[utoipa::path(
    post,
    path = "/pick-files",
    tag = "dialog",
    request_body(content = Option<PickFileRequest>, content_type = "application/json"),
    responses(
        (status = 200, description = "Multi-file selection result", body = PickFilesResponse),
    ),
)]
pub async fn pick_files(body: Option<Json<PickFileRequest>>) -> Json<PickFilesResponse> {
    let req = body.map(|Json(r)| r).unwrap_or_default();
    let dialog = build_file_dialog(
        req.title.as_deref(),
        "Select Files",
        req.filters.as_deref(),
        req.default_path.as_deref(),
    );
    let handles = dialog.pick_files().await;
    let paths = handles
        .unwrap_or_default()
        .iter()
        .map(|h| h.path().to_string_lossy().to_string())
        .collect();
    Json(PickFilesResponse { paths })
}

/// Build an `AsyncFileDialog` with optional title, filters, and default path.
fn build_file_dialog(
    title: Option<&str>,
    default_title: &str,
    filters: Option<&[FileFilter]>,
    default_path: Option<&str>,
) -> rfd::AsyncFileDialog {
    let mut dialog = rfd::AsyncFileDialog::new().set_title(title.unwrap_or(default_title));
    if let Some(filters) = filters {
        for f in filters {
            let ext_refs: Vec<&str> = f.extensions.iter().map(String::as_str).collect();
            dialog = dialog.add_filter(&f.name, &ext_refs);
        }
    }
    if let Some(path) = default_path {
        dialog = dialog.set_directory(path);
    }
    dialog
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn pick_file_request_deserialize_empty() {
        let json = "{}";
        let req: PickFileRequest = serde_json::from_str(json).unwrap();
        assert_eq!(req.title, None);
        assert_eq!(req.filters, None);
        assert_eq!(req.default_path, None);
    }

    #[test]
    fn pick_file_request_deserialize_full() {
        let json = r#"{
            "title": "Choose",
            "filters": [{"name": "Images", "extensions": ["png", "jpg"]}],
            "defaultPath": "/tmp"
        }"#;
        let req: PickFileRequest = serde_json::from_str(json).unwrap();
        assert_eq!(req.title.as_deref(), Some("Choose"));
        assert_eq!(req.filters.as_ref().unwrap().len(), 1);
        assert_eq!(req.filters.as_ref().unwrap()[0].name, "Images");
        assert_eq!(
            req.filters.as_ref().unwrap()[0].extensions,
            vec!["png", "jpg"]
        );
        assert_eq!(req.default_path.as_deref(), Some("/tmp"));
    }

    #[test]
    fn pick_folder_request_deserialize_empty() {
        let json = "{}";
        let req: PickFolderRequest = serde_json::from_str(json).unwrap();
        assert_eq!(req.title, None);
        assert_eq!(req.default_path, None);
    }

    #[test]
    fn pick_folder_request_deserialize_full() {
        let json = r#"{"title": "Pick Dir", "defaultPath": "/home"}"#;
        let req: PickFolderRequest = serde_json::from_str(json).unwrap();
        assert_eq!(req.title.as_deref(), Some("Pick Dir"));
        assert_eq!(req.default_path.as_deref(), Some("/home"));
    }

    #[test]
    fn pick_file_response_serialize_with_path() {
        let resp = PickFileResponse {
            path: Some("/tmp/file.txt".to_string()),
        };
        let json = serde_json::to_value(&resp).unwrap();
        assert_eq!(json["path"], "/tmp/file.txt");
    }

    #[test]
    fn pick_file_response_serialize_cancelled() {
        let resp = PickFileResponse { path: None };
        let json = serde_json::to_value(&resp).unwrap();
        assert!(json["path"].is_null());
    }

    #[test]
    fn pick_files_response_serialize_with_paths() {
        let resp = PickFilesResponse {
            paths: vec!["/a.txt".to_string(), "/b.txt".to_string()],
        };
        let json = serde_json::to_value(&resp).unwrap();
        assert_eq!(json["paths"], serde_json::json!(["/a.txt", "/b.txt"]));
    }

    #[test]
    fn pick_files_response_serialize_cancelled() {
        let resp = PickFilesResponse { paths: vec![] };
        let json = serde_json::to_value(&resp).unwrap();
        assert_eq!(json["paths"], serde_json::json!([]));
    }
}
